ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -356,8 +368,51 @@ void fillChunks() {
           if (chunkCounter != 0 && lastChunk.buf.position() == 0) {
             logger.trace("The last buffer(s) received from chunkFillerChannel have no data, discarding them.");
           } else {
-            lastChunk.onFillComplete(true);
-            updateChunkFillerWaitTimeMetrics();
+            if (!passedInBlobProperties.isEncrypted()) {
+              lastChunk.buf.flip();
+              lastChunk.onFillComplete(true);
+              updateChunkFillerWaitTimeMetrics();
+            } else {
+              try {
+                updateChunkFillerWaitTimeMetrics();
+                logger.trace("Chunk state moves to " + ChunkState.Encrypting);
+                lastChunk.state = ChunkState.Encrypting;
+                lastChunk.buf.flip();
+                lastChunk.chunkBlobSize = lastChunk.buf.remaining();
+                logger.trace("Submitting encrypt job for last chunk");
+                exec.submitJob(
+                    new EncryptJob(passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
+                        lastChunk.buf, ByteBuffer.wrap(lastChunk.chunkUserMetadata), kms.getRandomKey(), cryptoService,
+                        kms, (EncryptJob.EncryptJobResult result, Exception exception) -> {
+                      logger.trace("Processing encrypt job callback for last chunk");
+                      if (exception == null && !isOperationComplete()) {
+                        lastChunk.buf = result.getEncryptedBlobContent();
```

Comment:
`lastChunk.buf` gets reused (to prevent reallocations). I think the code relies on the fact that `buf` has a capacity of routerMaxChunkSize (and only alters the limits when the actual chunk is different). I would prefer that you leave this `buf` be and not overwrite it. Perhaps you should keep a separate variable for `encryptedChunkBuf` and `encryptedChunkMetadata` rather than overwriting them.

Reply:
Yeah, but to remember that this is the last chunk of this operation. I can fix things to not over-ride, but letting you that the reason you quoted may not be valid. But will the other place i.e. line 921 for intermittent chunks.

Reply:
Just now realized why I took this route. I don't want to maintain two different buffers, one to hold encryted  buf and other other non-encryption and only one of them will be non null. If we go this route, then at places where I need to fetch chunk's data, I have to do a ternary like isEncrypted()? encryptedBuf: buf sort of, which I am not a fan of. So, I wanted to have the same buffer hold the data (irrespective of whether its encrypted or not doesn't make any diff). Also, wrt over-riding, I did try to see if I can write to the same buffer which I couldnt. Bcoz, the encrypted data size exceeds the initial buffer size and hence. 
 A tangential point to note: Do keep in mind that buffers are re-used only within same Operation and not across operations. (if a blob is expected to be encrypted, then all of its chunks will be encrypted and hence we don't have to worry about buffer size mismatch between multiple chunks)